// react-lifecycle-visualizer/dist/components/Log.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactRedux = require("react-redux");

var constants = _interopRequireWildcard(require("../constants"));

var ActionCreators = _interopRequireWildcard(require("../redux/actionCreators"));

var _VisualizerProvider = require("../redux/VisualizerProvider");

var _LogEntries = _interopRequireDefault(require("./LogEntries"));

var _SimpleButton = _interopRequireDefault(require("./SimpleButton"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DelaySelector = function DelaySelector(_ref) {
  var value = _ref.value,
      onChange = _ref.onChange;
  return /*#__PURE__*/_react["default"].createElement("select", {
    value: value,
    onChange: onChange
  }, constants.delayValues.map(function (delay) {
    return /*#__PURE__*/_react["default"].createElement("option", {
      value: delay,
      key: delay
    }, delay, "s");
  }));
};

var Log = /*#__PURE__*/function (_Component) {
  _inherits(Log, _Component);

  var _super = _createSuper(Log);

  function Log() {
    var _this;

    _classCallCheck(this, Log);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "onKeyDown", function (evt) {
      if (evt.shiftKey) {
        // Require shift to prevent interference with scrolling
        switch (evt.code) {
          case 'ArrowUp':
            _this.props.stepLog(-1);

            break;

          case 'ArrowDown':
            _this.props.stepLog(1);

            break;
        }
      }
    });

    return _this;
  }

  _createClass(Log, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      document.addEventListener('keydown', this.onKeyDown);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      document.removeEventListener('keydown', this.onKeyDown);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return /*#__PURE__*/_react["default"].createElement("div", {
        className: "log"
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: "header"
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: "controls"
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: "title"
      }, "Log"), /*#__PURE__*/_react["default"].createElement("div", {
        className: "buttons"
      }, /*#__PURE__*/_react["default"].createElement(_SimpleButton["default"], {
        value: "clear log",
        onClick: function onClick() {
          return _this2.props.clearLog();
        }
      }), ' ', /*#__PURE__*/_react["default"].createElement("span", {
        className: "emoji-button",
        onClick: function onClick() {
          return _this2.props.stepLog(-1);
        }
      }, "\u23EA"), this.props.replayTimerId === null ? /*#__PURE__*/_react["default"].createElement("span", {
        className: "emoji-button",
        onClick: function onClick() {
          return _this2.props.startReplay();
        }
      }, "\u25B6\uFE0F") : /*#__PURE__*/_react["default"].createElement("span", {
        className: "emoji-button",
        onClick: function onClick() {
          return _this2.props.pauseReplay();
        }
      }, "\u23F8"), /*#__PURE__*/_react["default"].createElement("span", {
        className: "emoji-button",
        onClick: function onClick() {
          return _this2.props.stepLog(1);
        }
      }, "\u23E9")), /*#__PURE__*/_react["default"].createElement("div", null, "Delay:", ' ', /*#__PURE__*/_react["default"].createElement(DelaySelector, {
        value: this.props.replayTimerDelay,
        onChange: function onChange(evt) {
          return _this2.props.setDelay(+evt.currentTarget.value);
        }
      }))), /*#__PURE__*/_react["default"].createElement("div", {
        className: "hint"
      }, "(hover to highlight, shift-up/down to navigate)")), /*#__PURE__*/_react["default"].createElement(_LogEntries["default"], {
        entries: this.props.logEntries,
        highlightedIndex: this.props.highlightedIndex,
        highlight: this.props.highlight
      }));
    }
  }]);

  return Log;
}(_react.Component);

var mapStateToProps = function mapStateToProps(_ref2) {
  var logEntries = _ref2.logEntries,
      highlightedIndex = _ref2.highlightedIndex,
      replayTimerId = _ref2.replayTimerId,
      replayTimerDelay = _ref2.replayTimerDelay;
  return {
    logEntries: logEntries,
    highlightedIndex: highlightedIndex,
    replayTimerId: replayTimerId,
    replayTimerDelay: replayTimerDelay
  };
};

var _default = (0, _reactRedux.connect)(mapStateToProps, ActionCreators, null, {
  context: _VisualizerProvider.LifecycleVisualizerContext
})(Log);

exports["default"] = _default;
